api/controller.py

```diff
@@ -1031,20 +1036,25 @@ def fulfill(self, license_pool_id, mechanism_id=None, do_get=None):
 
         loan, loan_license_pool = self.get_patron_loan(patron, [pool])
 
-        if not loan or not loan_license_pool:
-            return NO_ACTIVE_LOAN.detailed(
-                _("You have no active loan for this title.")
-            )
+        requested_license_pool = loan_license_pool or pool
 
         # Find the LicensePoolDeliveryMechanism they asked for.
         mechanism = None
         if mechanism_id:
             mechanism = self.load_licensepooldelivery(
-                loan_license_pool, mechanism_id
+                requested_license_pool, mechanism_id
             )
             if isinstance(mechanism, ProblemDetail):
                 return mechanism
 
+        if (not loan or not loan_license_pool) and not (
+                self.can_fulfill_without_loan(
+                    library, patron, requested_license_pool, mechanism)
+        ):
+            return NO_ACTIVE_LOAN.detailed(
+                _("You have no active loan for this title.")
+            )
```

Comment:
When you can't fulfill without a loan but no patron is authenticated, is this the problem detail? It might be better to have a 401 in that case.